src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusExporterMiddleware.cs

```diff
@@ -67,34 +67,57 @@ public async Task InvokeAsync(HttpContext httpContext)
                 return;
             }
 
-            try
+            var buffer = new byte[65536];
+            var count = 0;
+
+            this.exporter.OnExport = (metrics) =>
             {
-                this.exporter.Collect(Timeout.Infinite);
+                try
+                {
+                    count = PrometheusSerializer.WriteMetrics(buffer, 0, metrics);
+                    return ExportResult.Success;
+                }
+                catch (Exception ex)
+                {
+                    PrometheusExporterEventSource.Log.FailedExport(ex);
+                    return ExportResult.Failure;
+                }
+            };
 
-                await WriteMetricsToResponse(this.exporter, response).ConfigureAwait(false);
+            try
+            {
+                if (this.exporter.Collect(Timeout.Infinite))
+                {
+                    await WriteMetricsToResponse(buffer, count, response).ConfigureAwait(false);
```

Comment:
@reyang I pushed an update to your branch I hope you don't mind. The PR was just updating `PrometheusExporterMetricsHttpServer` before. I updated `PrometheusExporterMiddleware` in my commit. It is the far more interesting case because, by default, you can't do a synchronous write to the AspNetCore response, it will throw an exception. Check out what I did here. I used a closure so we write to the buffer sync, and then we do an async write to the response. Should work and doesn't tie up two threads.